File: output/llama3.1-8b/avatar/Python/Java/atcoder_ABC150_E.java
import java.util.Arrays;
import java.util.Scanner;

public class atcoder_ABC150_E {
    public static void main(String[] args) {
        int mod = (int) Math.pow(10, 9) + 7;
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int[] c = new int[n];
        for (int i = 0; i < n; i++) {
            c[i] = scanner.nextInt();
        }
        Arrays.sort(c);
        reverseArray(c);

        long b = pow(2, 2 * n - 2, mod);
        long a = 2 * b % mod;
        int ans = 0;
        for (int i = 0; i < n; i++) {
            ans += c[i] * (a + i * b);
            ans %= mod;
        }
        System.out.println(ans);
    }

    public static void reverseArray(int[] array) {
        int left = 0, right = array.length - 1;
        while (left < right) {
            int temp = array[left];
            array[left] = array[right];
            array[right] = temp;
            left++;
            right--;
        }
    }

    public static long pow(long base, long exponent, int mod) {
        if (exponent == 0)
            return 1;
        else if (exponent % 2 == 0)
            return pow(base * base % mod, exponent / 2, mod);
        else
            return pow(base * base % mod, (exponent - 1) / 2, mod) * base % mod;
    }
}

File: output/llama3.1-8b/avatar/Java/Python/atcoder_ABC150_E.py
import sys
import math

def power(x, n):
    if n == 0:
        return 1
    mod = 10**9 + 7
    val = power(x, n // 2)
    val = (val * val) % mod
    if n % 2 == 1:
        val = (val * x) % mod
    return val

def main():
    sa = input().split()
    n = int(sa[0])
    c = list(map(int, input().split()))
    mod = 10**9 + 7
    if n == 1:
        print((c[0] * 2) % mod)
        return
    c.sort()
    b = power(2, n)
    a = power(2, n - 2)
    ans = 0
    for i in range(2, n + 2):
        val = (a * i) % mod
        val *= c[n - i]
        val %= mod
        ans += val
        ans %= mod
    ans *= b
    ans %= mod
    print(ans)

if __name__ == "__main__":
    main()